bpo-39939: Add str.removeprefix and str.removesuffix (#18939)
python/cpython: Doc/library/stdtypes.rst

   interpreted as in slice notation.

.. method:: str.removeprefix(prefix, /)

   If the string starts with the *prefix* string, return
   ``string[len(prefix):]``. Otherwise, return a copy of the original
   string::

      >>> 'TestHook'.removeprefix('Test')
      'Hook'
      >>> 'BaseTestCase'.removeprefix('Test')
      'BaseTestCase'

   .. versionadded:: 3.9

.. method:: str.removesuffix(suffix, /)

   If the string ends with the *suffix* string, return

>    ``string[:len(string)-len(suffix)]``. Otherwise, return a copy of the

Reviewer: Why not removing len(string)?

```suggestion
   ``string[:-len(suffix)]``. Otherwise, return a copy of the
```

Author: Just removing it would create a an issue with ``string[:-0]`` when the ``suffix`` is empty, so I'll change it to
```
   If the string ends with the *suffix* string and the *suffix* is not empty, return
   ``string[:-len(suffix)]``. Otherwise, return a copy of the
```

Reviewer: Alright. Adding " and the *suffix* is not empty" sounds reasonable.

   interpreted as in slice notation.

.. method:: str.removeprefix(prefix, /)

> 

Reviewer: Oh, the docstring definition is short and sounds correct:

```
If the string starts with the *prefix* string, return `str[len(prefix):]`.
```

This definition implicitly documents that the prefix is removed at most once.

>    Return a copy of the string with the given prefix removed, if present. ::

Reviewer: ```suggestion
   Return a copy of the string with the given *prefix* string removed, if present. ::
```

>       >>> 'BarFooBaz'.removeprefix('Bar')
>       'FooBaz'
>       >>> 'BarFooBaz'.removeprefix('Baz')
>       'BarFooBaz'

Reviewer: These examples are not easy to read since Bar and Baz look the same. I suggest to use different strings:

```suggestion
      >>> 'TestHook'.removeprefix('Test')
      'Hook'
      >>> 'BaseTestCase'.removeprefix('Test')
      'BaseTestCase'
```

   The expression ``s.removeprefix(pre)`` is roughly equivalent to
   ``s[len(pre):] if s.startswith(pre) else s``.
   Unlike :meth:`~str.startswith`, only one prefix can be passed

>    at a time.

Reviewer: I don't think that it's worth it to mention that. Same remark for removesuffix() doc.

> 

Reviewer: You should mention that even if the prefix is present twice or more, it is only removed once.

   .. versionadded:: 3.9

.. method:: str.removesuffix(suffix, /)

>    Return a copy of the string with the given suffix removed, if present. ::

Reviewer: ```suggestion
   Return a copy of the string with the given *suffix* string removed, if present. ::
```

>       >>> 'BarFooBaz'.removesuffix('Baz')
>       'BarFoo'
>       >>> 'BarFooBaz'.removesuffix('Bar')
>       'BarFooBaz'

Reviewer: ```suggestion
      >>> 'MiscTests'.removesuffix('Tests')
      'Misc'
      >>> 'TmpDirMixin'.removesuffix('Tests')
      'TmpDirMixin'
```

      Also accept an integer in the range 0 to 255 as the subsequence.

.. method:: bytes.removeprefix(prefix, /)
            bytearray.removeprefix(prefix, /)

>    If the binary data starts with the *prefix*, return ``b[len(prefix):]``.

Reviewer: ```suggestion
   If the binary data starts with the *prefix* string, return ``bytes[len(prefix):]``.
```
